### PR TITLE
Fix Duplicate Dependencies Showing in PR summary table

### PR DESCRIPTION
**Pull request by honeyankit**
## Context

This PR resolves the issue of duplicated entries in the group update PR's summary table by ensuring that only unique dependencies are considered in the message builder. These are then displayed in the PR summary table.

Fixes: https://github.com/dependabot/dependabot-core/issues/7695

### CODE REVIEW

File: common/lib/dependabot/pull_request_creator/message_builder.rb [hunk 1]
```diff
@@ -498,13 +498,15 @@ def multi_directory_group_intro
       sig { returns(String) }
       def group_intro
         update_count = dependencies.map(&:name).uniq.count
+        # Ensure dependencies are unique by name
+        unique_dependencies = dependencies.uniq { |dep| [dep.name] }
```
Comment (jurre) [inline common/lib/dependabot/pull_request_creator/message_builder.rb:502]:
```suggestion
unique_dependencies = dependencies.uniq(&:name)
```

**Comment by jurre** [comment]
Can we add a test to prevent regression?

**Comment by honeyankit** [comment]
> Can we add a test to prevent regression?

I have tested against the test repo. I am adding the test.

**Comment by Nishnha** [comment]
There's also a mention of the table showing an update from a commit to a version, which I think is why they're showing up even though they didn't update: https://github.com/dependabot/dependabot-core/issues/7695#issuecomment-1662267570

Is a fix for that going to be part of this PR?

**Comment by honeyankit** [comment]
> There's also a mention of the table showing a dependency updating from a commit to a version, even though they didn't actually update: [#7695 (comment)](https://github.com/dependabot/dependabot-core/issues/7695#issuecomment-1662267570)
> 
> Is a fix for that going to be part of this PR?

No. This will fix the duplicate entry. I can look into that as well. But first will let this ship

**Review by landongrindheim — APPROVED** [review]

File: common/spec/dependabot/pull_request_creator/message_builder_spec.rb [hunk 2]
```diff
@@ -2309,6 +2309,116 @@ def commits_details(base:, head:)
...
+            Dependabot::Dependency.new(
+              name: "business5",
+              version: "0.17.0",
+              previous_version: "0.16.2",
+              package_manager: "dummy",
+              requirements: [],
+              previous_requirements: []
+            )
+          end
+          let(:dependencies) { [dependency, dependency2, dependency3, dependency4, dependency5, dependency6] }
```
[the remark this answers is not in the document]
Comment (honeyankit) [reply common/spec/dependabot/pull_request_creator/message_builder_spec.rb:2363]: I like this. Adding it now

**Comment by Nishnha** [comment]
Since we're only uniq'ing by name, wouldn't the dependencies that gets removed from the PR body be random? I wonder if the "to" and "from" version can change for a duplicate dependency if you run the same job twice

**Comment by honeyankit** [comment]
@Nishnha : For less then 5 dependencies, similar approach is used [here](https://github.com/dependabot/dependabot-core/blob/005dd7314eb6f4dfc2f1505631f01e19f3138c73/common/lib/dependabot/pull_request_creator/message_builder.rb#L610-L615) where by the dependency name the unique dependencies are filtered. So I have just refer that for more then 5 dependencies.

**Comment by Nishnha** [comment]
> @Nishnha : For less then 5 dependencies for group PR, similar approach is used [here](https://github.com/dependabot/dependabot-core/blob/005dd7314eb6f4dfc2f1505631f01e19f3138c73/common/lib/dependabot/pull_request_creator/message_builder.rb#L610-L615) where by the dependency name the unique dependencies are filtered. So I have just refer that for more then 5 dependencies.

Right but the dependency version isn't displayed for <5 dependencies

**Comment by jurre** [comment]
We could check for unique by name, from and to to be thorough?

**Comment by honeyankit** [comment]
@jurre and @Nishnha : I have updated the code to remove duplicates based on name, to and from versions. 

1. Both Dep A will be shown
```bash
Dep A - previous version 1.2.4 latest version 1.2.5 
Dep A - previous version 1.2.4 latest version 1.2.6 
```

2. Both Dep A will be shown
```bash
Dep A - previous version 1.2.1 latest version 1.2.5 
Dep A - previous version 1.2.4 latest version 1.2.5
```

**Comment by honeyankit** [comment]
All the smoke test errors are due to rate limit exceeded.

File: common/spec/dependabot/pull_request_creator/message_builder_spec.rb [hunk 3]
```diff
@@ -2422,6 +2422,142 @@ def commits_details(base:, head:)
           end
         end
 
+        context "with five or more dependencies with same name, differet to and from versions", :vcr do
```
Comment (Nishnha) [inline common/spec/dependabot/pull_request_creator/message_builder_spec.rb]: this line has a type for different
do we need the vcr? I don't see one added here

Comment (honeyankit) [reply common/spec/dependabot/pull_request_creator/message_builder_spec.rb]: Removed all the unnecessary vcr from all the tests. I had copied it from other test which already had the vcr along with manual stub. Good catch.

Comment (honeyankit) [reply common/spec/dependabot/pull_request_creator/message_builder_spec.rb]: Typo was fixed already: https://github.com/dependabot/dependabot-core/pull/9436/commits/4d241b2d02d95988a25f8e1d248bb1080edc3d92

**Review by Nishnha — APPROVED** [review]
Nice

**Comment by edmorley** [comment]
@honeyankit @Nishnha @landongrindheim Could this have caused the regression seen in #9457?

**Pull request merged**